fabianlim/fms-hf-tuning
# Acceleration Framework

### Description of the change

This PR will include the following features:
- implement an `AccelerationFramework` and integrate into `tuning/sft_trainer.py`
- implement the following `AccelerationFramework` plugins:
   * `AutoGPTQAccelerationPlugin`: load an already quantized checkpoint with the AutoGPTQ.
   * `BNBAccelerationPlugin`: load a checkpoint with HF `AutoModel.from_pretrained` loader  via the `BitsAndBytesConfig`.
   * `UnslothStandAloneAccelerationPlugin`: use the `unsloth` standalone loader, and patch an `auto_gptq` tritonv2. Currently only supports `auto_gptq`.
   * `UnslothStackableAccelerationPlugin`: extract out `unsloth` fast_lora, loss, rms, etc and install onto of a model loaded by standard loaders (e.g, `AutoGPTQAccelerationPlugin` and `BNBAccelerationPlugin`)
 
**Notes**:
- There is a issue with casting, as mentioned in the [unsloth blog, item 1](https://unsloth.ai/blog/mistral-benchmark), to fix this, we have to omit an fp32 upcast in `prepare_model_for_kbit_training`.
- We discovered that release [40.0.0](https://github.com/huggingface/transformers/releases/tag/v4.40.0) broke the speed of the auto_gptq model. ~This is probably due to a type issue.~ 
   * Found the breaking commit to be https://github.com/huggingface/transformers/commit/40eb6d6c5fcdaf99b499c249640d41f28659565b
    * turns out that mixtral's SDPA implementation was not using a windowed attention up until this commit (due to a 4d mask problem)
    * @achew010 raised a HF issue https://github.com/huggingface/transformers/issues/30461
    * See the updated explaination [here](https://github.com/fabianlim/fms-hf-tuning/pull/4#issuecomment-2073973523).
    * The unsloth exprs should be correct, since it will create [an xformers windowed mask](https://github.com/unslothai/unsloth/blob/ec19e61c854dcf9104386fa63fc6c4f2944d4f35/unsloth/models/mistral.py#L190-L192) and [pass the mask to SDPA](https://github.com/unslothai/unsloth/blob/ec19e61c854dcf9104386fa63fc6c4f2944d4f35/unsloth/models/mistral.py#L153).  
    *  FA2 will not be affected since it handles the masking internally in the kernel.
- Somme AutoGPTQ checkpoints (e.g., `TheBloke/Mistral-7B-v0.1-GPTQ`) are known to not have bias config, but we discovered that the checkpoint files have empty bias entries.

**Limitations**
- `AutoGPTQAccelerationPlugin `: currently only supports pre-quantized checkpoints.
- `unsloth` currently only works if mixed precision is turned on.
- `package_requirement` checks are currently broken.

### Related issue number



### How to verify the PR 



Run the script as usual, but (see example below):
- specify the `--acceleration_framework_config_file` key pointing to the above yaml file
- ensure that a GTPQ'ed checkpoint is used, e.g., [`TheBloke/Mixtral-8x7B-Instruct-v0.1-GPTQ`](https://huggingface.co/TheBloke/Mixtral-8x7B-Instruct-v0.1-GPTQ).
- remember to ensure appropriate `dytpe`'s that the kernel supports (`triton_v2` kernel supports `float16` only) : 
    1. specifying the native dtype via `--torch_dtype`, or 
    2. specifying AMP (e.g., via `--fp16`). 
- specifying the `--peft_method lora` and the other related settings. 

```shell
python \
    tuning/sft_trainer.py \
    --training_data_path $DATA_PATH \
    --model_name_or_path mistralai/Mixtral-8x7B-Instruct-v0.1 \
    --acceleration_framework_config_file fixtures/acceleration_framework.yaml \
    --torch_dtype float16 \
    ...
    --peft_method lora \
	--r 16 --lora_alpha 16 --lora_dropout 0.1 \
    --target_modules q_proj k_proj v_proj o_proj 
```


### Was the PR tested

#### Packing Experiments (A100)

In this experiments, we tested a simple version of packing that does not have `attention_masks` properly set:
- loaded the `alpaca` dataset.
- turn on `packing=True` in `SFTTrainer`.
- pass `formatting_func`  into `SFTTrainer`.

Model | Base Model Size | Base Quant | Attn | toks/s (vanilla) *  | toks/s (with fused kernels, etc) 
--|--|--|--|--|--
mistralai/Mistral-7B-v0.1 | 7b | bnb-nf4 | SDPA | 3355 | 4018
TheBloke/Mistral-7B-v0.1-GPTQ | 7b  | auto_gptq | SDPA | 3438 | 4104



*The Vanilla benchs are running on transformers older than 4.40, which do not have the sliding mask properly set.* 

![image](https://github.com/fabianlim/fms-hf-tuning/assets/8325951/01eda2d5-c18d-43a7-ab83-ad0d9623d186)

#### Experiments Showing How Packing Affects Throughputs (A100)

Here are the packing comparisons for MiStral and MiXtral with/without stackable unsloth kernels

| Model  | Base Model Size | Base Quant | Packing | toks/s <br> (vanilla) | toks/s <br> (with fused kernels, etc) | % improvement | GPU
| -------- | ------- | ------- | ------- | ------- | ------- | ------- |--------|
| TheBloke/Mixtral-8x7B-Instruct-v0.1-GPTQ | 47b | auto_gptq | True | 2044 | 2079 | 1.7 | 1 |
| TheBloke/Mixtral-8x7B-Instruct-v0.1-GPTQ | 47b | auto_gptq | False | 668 | 669 | 0.0 | 1 |
| TheBloke/Mistral-7B-v0.1-GPTQ | 7b | auto_gptq | True | 3499 | 4166 | 19.0 | 1 |
| TheBloke/Mistral-7B-v0.1-GPTQ | 7b | auto_gptq | False | 2228 | 2375 | 6.6 | 1 |
| TheBloke/Llama-2-70B-GPTQ | 70b | auto_gptq | True | 452 | 502 | 11.1 | 1 |
| TheBloke/Llama-2-70B-GPTQ | 70b | auto_gptq | False | 334 | 366 | 9.5 | 1 |
| TheBloke/Llama-2-70B-GPTQ | 70b | auto_gptq | True | 440/device | 471/device (Without Fused Modules) <br><br> 486/device (NoFSDPWrappedFused) | 7 <br><br> | 2 |
| TheBloke/Llama-2-70B-GPTQ | 70b | auto_gptq | False | 93/device | 176/device (NoFSDPWrappedFused) | 89.2 | 2 |
| TheBloke/Llama-3-70B-GPTQ | 70b | auto_gptq | True | 405/device | 440/device (Without Fused Modules) <br><br> 454/device (NoFSDPWrappedFused) | 8.6 <br><br> | 2 |
| TheBloke/Llama-3-70B-GPTQ | 70b | auto_gptq | False | 70/device | 139/device (NoFSDPWrappedFused) | 98.5 | 2 |

*The Vanilla benchs are running on transformers older than 4.40, which do not have the sliding mask properly set.*

#### Experiments on L40

| Model  | Base Model Size | Base Quant | Packing | toks/s <br> (vanilla) | toks/s <br> (with fused kernels, etc) | Improvement
| -------- | ------- | ------- | ------- | ------- | ------- |--|
| TheBloke/Mixtral-8x7B-Instruct-v0.1-GPTQ | 47b | auto_gptq | True | 1484  | 1662  |  12 %
| TheBloke/Mistral-7B-v0.1-GPTQ | 7b | auto_gptq | True | 2488  | 3016 | 21%


* The Mixtral experimentsa are running on FA2 *
*The Vanilla benchs are running on transformers older than 4.40, which do not have the sliding mask properly set.*
